### PR TITLE
Fix issued that caused epoch to not cleanup

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Signed-off-by: Alan Jowett <alan.jowett@microsoft.com>

## Description
Cleanup the handling of flags in epoch code (put them all under the lock).
Set thread affinity on epoch_enter/exit (some paths where missing this).
Unify CPU and thread ebpf_epoch_state_t.
Fix issue with re-arming timer that caused use after free.

## Testing

Existing CI/CD tests.

## Documentation

Not needed.

### CODE REVIEW

File: libs/platform/ebpf_epoch.c [hunk 1]
```diff
@@ -40,24 +40,37 @@
 // Delay after the _ebpf_flush_timer is set before it runs.
 #define EBPF_EPOCH_FLUSH_DELAY_IN_MICROSECONDS 1000
 
+// Time before logging that a thread entry is stale
+#define EBPF_EPOCH_STALE_THREAD_TIME_IN_NANO_SECONDS 10000000000 // 10 second
```
Comment (dthaler) [inline libs/platform/ebpf_epoch.c:44]:
```suggestion
#define EBPF_EPOCH_STALE_THREAD_TIME_IN_NANO_SECONDS 10000000000 // 10 seconds
```

File: libs/platform/ebpf_epoch.c [hunk 2]
```diff
@@ -40,24 +40,37 @@
...
-    int32_t flags;
     ebpf_non_preemptible_work_item_t* stale_worker;
+    uint32_t padding;
 } ebpf_epoch_cpu_entry_t;
 
+typedef struct _ebpf_epoch_thread_entry
+{
+    ebpf_epoch_state_t epoch_state;
+    uintptr_t old_thread_affinity_mask;
+    uint64_t last_used;
```
Comment (dthaler) [inline libs/platform/ebpf_epoch.c:71]: It was hard for me to understand what this field meant.   Maybe
```suggestion
    uint64_t last_used_time;
```

File: libs/platform/ebpf_epoch.c [hunk 3]
```diff
@@ -576,66 +618,24 @@ _ebpf_flush_worker(_In_ void* context)
     ebpf_epoch_flush();
 }
 
-ebpf_result_t
-_ebpf_epoch_update_thread_state(uint32_t cpu_id, uintptr_t thread_id, int64_t current_epoch, bool enter)
+_Requires_lock_held_(_ebpf_epoch_cpu_table[cpu_id].lock) ebpf_epoch_thread_entry_t* _ebpf_epoch_get_thread_entry(
```
Comment (dthaler) [inline libs/platform/ebpf_epoch.c:621]: this function starts with an `_` and so should be marked static

File: libs/platform/unit/platform_unit_test.cpp [hunk 4]
```diff
@@ -327,36 +327,43 @@ TEST_CASE("epoch_test_stale_items", "[platform]")
...
-
-    std::thread thread_1(t1);
-    std::thread thread_2(t2);
-
-    thread_1.join();
-    thread_2.join();
-    std::this_thread::sleep_for(std::chrono::seconds(1));
-    REQUIRE(ebpf_epoch_is_free_list_empty(0));
-    REQUIRE(ebpf_epoch_is_free_list_empty(1));
+    for (size_t i = 0; i < 100; i++) {
```
Comment (dthaler) [inline libs/platform/unit/platform_unit_test.cpp:330]: if 100 arbitrary?   maybe
```suggestion
    const size_t test_iterations = 100; 
    for (size_t i = 0; i < test_iterations; i++) {
```

**Review by dthaler — APPROVED** [review]

File: libs/platform/unit/platform_unit_test.cpp (hunk 4 above)
Comment (dthaler) [reply libs/platform/unit/platform_unit_test.cpp]: I didn't see any response or change for this comment

**Review by dthaler — DISMISSED** [review]

File: libs/platform/unit/platform_unit_test.cpp (hunk 4 above)
Comment (Alan-Jowett) [reply libs/platform/unit/platform_unit_test.cpp]: Isn't it part of this commit https://github.com/microsoft/ebpf-for-windows/pull/1332/commits/3ea9b0ae272ba236ce2aadbf2f506a140b6ec270?

Comment (dthaler) [reply libs/platform/unit/platform_unit_test.cpp]: sort of.  I was looking for some label on the max.  the commit labeled the counter.   If you think it's good enough, feel free to ignore this comment.

File: libs/platform/ebpf_epoch.c [hunk 5]
```diff
@@ -40,24 +40,37 @@
 // Delay after the _ebpf_flush_timer is set before it runs.
 #define EBPF_EPOCH_FLUSH_DELAY_IN_MICROSECONDS 1000
 
+// Time before logging that a thread entry is stale
```
Comment (shankarseal) [inline libs/platform/ebpf_epoch.c:43]: Consider updating the summary at the top of the file with the new epoch states.

File: libs/platform/ebpf_epoch.c [hunk 6]
```diff
@@ -248,45 +237,81 @@ ebpf_epoch_terminate()
...
     if (current_cpu >= _ebpf_epoch_cpu_count) {
         return EBPF_OPERATION_NOT_SUPPORTED;
     }
 
+    ebpf_lock_state_t state = ebpf_lock_lock(&_ebpf_epoch_cpu_table[current_cpu].lock);
+
     if (ebpf_is_preemptible()) {
-        return _ebpf_epoch_update_thread_state(current_cpu, ebpf_get_current_thread_id(), _ebpf_current_epoch, true);
+        ebpf_epoch_thread_entry_t* thread_entry =
+            _ebpf_epoch_get_thread_entry(current_cpu, ebpf_get_current_thread_id(), true);
```
Comment (shankarseal) [inline libs/platform/ebpf_epoch.c:253]: nit: to make this more readable can we use a comment describing the last bool param or use a constant identifier instead?

File: libs/platform/ebpf_epoch.c [hunk 7]
```diff
@@ -40,24 +40,37 @@
...
-    _Requires_lock_held_(lock) ebpf_epoch_state_t cpu_epoch_state;
+    _Requires_lock_held_(lock) ebpf_epoch_state_t epoch_state;
     _Requires_lock_held_(lock) ebpf_list_entry_t free_list;
     _Requires_lock_held_(lock) ebpf_hash_table_t* thread_table;
-    int32_t flags;
     ebpf_non_preemptible_work_item_t* stale_worker;
+    uint32_t padding;
 } ebpf_epoch_cpu_entry_t;
 
+typedef struct _ebpf_epoch_thread_entry
```
Comment (shankarseal) [inline libs/platform/ebpf_epoch.c:67]: description of various fields?

File: libs/platform/ebpf_epoch.c [hunk 8]
```diff
@@ -520,13 +545,22 @@ _ebpf_epoch_get_release_epoch(_Out_ int64_t* release_epoch)
...
                 }
             } else {
-                _ebpf_set_per_cpu_flag(&_ebpf_epoch_cpu_table[cpu_id], EBPF_EPOCH_PER_CPU_STALE, true);
+                _ebpf_epoch_cpu_table[cpu_id].epoch_state.stale = true;
+            }
+            if (!_ebpf_epoch_cpu_table[cpu_id].epoch_state.timer_armed &&
+                !_ebpf_epoch_cpu_table[cpu_id].epoch_state.shutdown) {
+
+                _ebpf_epoch_cpu_table[cpu_id].epoch_state.timer_armed = true;
+
```
Comment (shankarseal) [inline libs/platform/ebpf_epoch.c:554]: suggestion: keeping `stale_items` local variable and scheduling the timer at the end of the function might reduce the code delta a bit.

File: libs/platform/unit/platform_unit_test.cpp (hunk 4 above)
Comment (Alan-Jowett) [reply libs/platform/unit/platform_unit_test.cpp]: ok, done.

File: libs/platform/ebpf_epoch.c [hunk 9]
```diff
@@ -36,29 +36,46 @@
 //
 // _ebpf_epoch_stale_worker:
 // Called if a CPU has items to be freed, but hasn't run anything in EBPF_EPOCH_FLUSH_DELAY_IN_MICROSECONDS.
+//
+// Stale flag:
+// The stale flag is set if the timer runs and the ebpf_epoch_cpu_entry_t has entries in it's free list.
```
Comment (dthaler) [inline libs/platform/ebpf_epoch.c:41]:
```suggestion
// The stale flag is set if the timer runs and the ebpf_epoch_cpu_entry_t has entries in its free list.
```
fix grammar

File: libs/platform/ebpf_epoch.c [hunk 10]
```diff
@@ -112,49 +130,65 @@ typedef struct _ebpf_epoch_work_item
...
-        ebpf_interlocked_or_int32(&cpu_entry->flags, flag);
-    } else {
-        ebpf_interlocked_and_int32(&cpu_entry->flags, ~flag);
-    }
-}
+/**
+ * @brief Arm the flush timer if timer is:
+ *  Not already armed.
+ *  Not disabled.
+ *  Free list is not empty.
```
Comment (dthaler) [inline libs/platform/ebpf_epoch.c:189]:
```suggestion
 * @brief Arm the flush timer if:
 *  Timer is not already armed.
 *  Timer is not disabled.
 *  Free list is not empty.
```
fix grammar

File: libs/platform/ebpf_epoch.c [hunk 11]
```diff
@@ -520,13 +601,13 @@ _ebpf_epoch_get_release_epoch(_Out_ int64_t* release_epoch)
...
             } else {
-                _ebpf_set_per_cpu_flag(&_ebpf_epoch_cpu_table[cpu_id], EBPF_EPOCH_PER_CPU_STALE, true);
+                _ebpf_epoch_cpu_table[cpu_id].epoch_state.stale = true;
             }
-            stale_items = true;
         }
 
-        if (_ebpf_epoch_cpu_table[cpu_id].cpu_epoch_state.active) {
-            lowest_epoch = min(lowest_epoch, _ebpf_epoch_cpu_table[cpu_id].cpu_epoch_state.epoch);
+        // Include ths epoch state if it's active.
```
Comment (dthaler) [inline libs/platform/ebpf_epoch.c:608]:
```suggestion
        // Include this epoch state if it's active.
```

File: libs/platform/ebpf_epoch.c [hunk 12]
```diff
@@ -536,16 +617,31 @@ _ebpf_epoch_get_release_epoch(_Out_ int64_t* release_epoch)
...
+                (uint8_t**)&thread_entry);
 
+            // There are no more entries in the table.
             if (return_value != EBPF_SUCCESS) {
                 break;
             }
-            if (thread_epoch_state->active) {
-                lowest_epoch = min(lowest_epoch, thread_epoch_state->epoch);
+
+            // Include ths epoch state if it's active.
```
Comment (dthaler) [inline libs/platform/ebpf_epoch.c:627]:
```suggestion
            // Include this epoch state if it's active.
```

**Review by dthaler — APPROVED** [review]
Ok modulo typos/grammar

**Review by dthaler — APPROVED** [review]

**Review by shankarseal — APPROVED** [review]

**Pull request merged**